oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp

```diff
@@ -0,0 +1,280 @@
+                              PAM_SILENT | PAM_DISALLOW_NULL_AUTHTOK);
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        return;
+    }
+
+    authenticated = true;
+
+    /* check that the account is healthy */
+    retval = pam_acct_mgmt(pamHandle, PAM_DISALLOW_NULL_AUTHTOK);
+    if (retval == PAM_NEW_AUTHTOK_REQD)
+    {
+        operationAllowed = false;
+        passwordChangeRequired = true;
+        passwordChangeAllowed = true;
+        pam_end(pamHandle, retval); // ignore retval
+        return;
+    }
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        return;                     // operationAllowed == false
+    }
+    operationAllowed = true;
+    passwordChangeAllowed = true;
+
+    retval = pam_end(pamHandle, retval); // ignore retval
+}
+
+void authenticate(const std::string& username, const std::string& password,
+                  bool& authenticated, bool& passwordChangeRequired,
+                  bool& operationAllowed)
+{
+    bool passwordChangeAllowed = false;
+    common_authentication_handler(username, password, authenticated,
+                                  passwordChangeRequired, operationAllowed,
+                                  passwordChangeAllowed);
+}
+
+enum changePasswordReasonCode changePassword(const std::string& username,
+                                             const std::string& currentPassword,
+                                             const std::string& newPassword,
+                                             bool& operationAllowed)
+{
+    bool passwordChangeRequired = false;
+    operationAllowed = false;
+    bool authenticated = false;
+    bool passwordChangeAllowed = false;
+    common_authentication_handler(username, currentPassword, authenticated,
+                                  passwordChangeRequired, operationAllowed,
+                                  passwordChangeAllowed);
+    if (!authenticated)
+    {
+        return NOT_AUTHENTICATED;
+    }
+    if (!passwordChangeAllowed)
+    {
+        return NOT_ALLOWED;
+    }
+
+    struct pfw_sms_pam_appdata appdata = {
+        newPassword.c_str(), PASSWORD_CHANGE_FAILED_UNKNOWN_REASON};
+    char* appData = reinterpret_cast<char*>(&appdata);
+    const struct pam_conv pamConversation = {pamConversationFunction, appData};
+    pam_handle_t* pamHandle = nullptr; // this gets set by pam_start
+    int retval = pam_start("ibm-pfw-sms-menu", username.c_str(),
+                           &pamConversation, &pamHandle);
+    if (retval != PAM_SUCCESS)
+    {
+        return PAM_START_FAILED;
+    }
+
+    retval = pam_chauthtok(pamHandle, PAM_SILENT);
+    if (retval != PAM_SUCCESS)
+    {
+        pam_end(pamHandle, retval); // ignore retval
+        // Return the specific reason gathered from the conversation function
+        return appdata.reasonCode;
+    }
+
+    retval = pam_end(pamHandle, retval); // ignore retval
```

Comment:
assigning of retval won't make any sense since it's last line and retval is local scope & function returns nothing.

Reply:
Same comment as above (at a different place in the code).  I will remove the assignment.

And note it is customary practice for Linux-PAM clients to ignore the return value from pam_end() when the previous operations have all been successful.  For example: https://github.com/linux-pam/linux-pam/blob/master/examples/vpass.c